From 3cbd094da4196e3142e4c9dec1d21db0ad865b96 Mon Sep 17 00:00:00 2001
From: Xiao Chen <xiao@apache.org>
Date: Wed, 15 Nov 2017 16:43:25 -0800
Subject: [PATCH 2678/2748] HADOOP-15023. ValueQueue should also validate
 (int) (lowWatermark * numValues) > 0 on
 construction.

(cherry picked from commit b1941b200d6b4fa6a7891421c0a1e212cad3d6eb)
(cherry picked from commit ba2eab9122f8c5b279c657f4750c74accb9a6407)

Change-Id: Ica3f69b65fc0c284bd51a29f8e82d5c2b376522c
---
 .../apache/hadoop/crypto/key/kms/ValueQueue.java   |    6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/crypto/key/kms/ValueQueue.java b/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/crypto/key/kms/ValueQueue.java
index 9e3c2e3..7ffed40 100644
--- a/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/crypto/key/kms/ValueQueue.java
+++ b/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/crypto/key/kms/ValueQueue.java
@@ -224,6 +224,9 @@ public ValueQueue(final int numValues, final float lowWatermark,
     Preconditions.checkArgument(numValues > 0, "\"numValues\" must be > 0");
     Preconditions.checkArgument(((lowWatermark > 0)&&(lowWatermark <= 1)),
         "\"lowWatermark\" must be > 0 and <= 1");
+    final int watermarkValue = (int) (numValues * lowWatermark);
+    Preconditions.checkArgument(watermarkValue > 0,
+        "(int) (\"numValues\" * \"lowWatermark\") must be > 0");
     Preconditions.checkArgument(expiry > 0, "\"expiry\" must be > 0");
     Preconditions.checkArgument(numFillerThreads > 0,
         "\"numFillerThreads\" must be > 0");
@@ -243,8 +246,7 @@ public ValueQueue(final int numValues, final float lowWatermark,
                       throws Exception {
                     LinkedBlockingQueue<E> keyQueue =
                         new LinkedBlockingQueue<E>();
-                    refiller.fillQueueForKey(keyName, keyQueue,
-                        (int)(lowWatermark * numValues));
+                    refiller.fillQueueForKey(keyName, keyQueue, watermarkValue);
                     return keyQueue;
                   }
                 });
-- 
1.7.9.5

